janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/step/JanusGraphPropertiesStep.java

```diff
@@ -82,55 +95,28 @@ public void setUseMultiQuery(boolean useMultiQuery) {
         return (Iterator<E>) Iterators.transform(iterable.iterator(), Property::value);
     }
 
-    private void initialize() {
-        assert !initialized;
-        initialized = true;
-        assert getReturnType().forProperties() || (orders.isEmpty() && hasContainers.isEmpty());
-
-        if (!starts.hasNext()) throw FastNoSuchElementException.instance();
-        final List<Traverser.Admin<Element>> elements = new ArrayList<>();
-        starts.forEachRemaining(elements::add);
-        starts.add(elements.iterator());
-        assert elements.size() > 0;
-
-        useMultiQuery = useMultiQuery && elements.stream().allMatch(e -> e.get() instanceof Vertex);
-
-        if (useMultiQuery) {
-            initializeMultiQuery(elements);
-        }
-    }
-
     /**
-     * This initialisation method is called the first time this instance is used and also when
-     * an attempt to retrieve a vertex from the cached multiQuery results doesn't find an entry.
-     * @param vertices A list of vertices with which to initialise the multiQuery
+     * This initialisation method is called when an attempt to retrieve a vertex from the cached multiQuery results
+     * doesn't find an entry.
      */
-    private void initializeMultiQuery(final List<Traverser.Admin<Element>> vertices) {
-        assert vertices.size() > 0;
-        final JanusGraphMultiVertexQuery multiQuery = JanusGraphTraversalUtil.getTx(traversal).multiQuery();
-        vertices.forEach(v -> multiQuery.addVertex((Vertex)v.get()));
+    private void prefetchNextBatch() {
+        final JanusGraphMultiVertexQuery multiQuery = JanusGraphTraversalUtil.getTx(getTraversal()).multiQuery();
+        multiQuery.addAllVertices(verticesToPrefetch);
+        verticesToPrefetch.clear();
         makeQuery(multiQuery);
 
-        Map<JanusGraphVertex, Iterable<? extends JanusGraphProperty>> results = multiQuery.properties();
-        if (multiQueryResults == null) {
-            multiQueryResults = results;
-        } else {
-            multiQueryResults.putAll(results);
+        try{
+            multiQueryResults = multiQuery.properties();
+        } catch (JanusGraphException janusGraphException) {
+            throw new TraversalInterruptedException();
         }
-        initialized = true;
-    }
-
-    @Override
-    protected Traverser.Admin<E> processNextStart() {
-        if (!initialized) initialize();
-        return super.processNextStart();
     }
 
     @Override
     protected Iterator<E> flatMap(final Traverser.Admin<Element> traverser) {
-        if (useMultiQuery) { //it is guaranteed that all elements are vertices
+        if (useMultiQuery && traverser.get() instanceof Vertex) {
```

Comment:
An equivalent check was done before in the old `initialize()` method of this class. That method aggregated all traversers, checked them one by one and only then decided whether to use MultiQuery or not.
Since the new approach should not require to fetch _all_ traversers, I had to generalize the type parameter of `JanusGraphMultiQueryStep` from `Vertex` to `Element` because we can not be sure that all elements are actually vertices before we have seen them all.